slack_wrapper.gs

```diff
@@ -0,0 +1,184 @@
+      output.msg = 'error: VERIFICATION_TOKEN is not set in script properties. The command will not run. Please contact the web app developer.';
+      return(output);
+    }  
+    if (this.token !== token_true) {
+      output.msg = 'error: Invalid token ' + this.token + ' . The command will not run. Please contact the web app developer.';
+      return(output);
+    }
+    
+    // check request originates from our slack workspace
+    if (this.teamid != teamid_true){
+      output.msg = 'error: You are sending your command from an unauthorised slack workspace.';
+      return(output);
+    }
+    
+    output.code=true;
+    return (output);
+  }
+  
+  
+  checkArgSyntax(){
+    //**** todo ****//
+  }
+  
+  handleEvent (){
+    //  handle slack doPOST events
+  
+    // extract relevant data from message body
+    var workspace = this.teamid;
+    var command = this.subtype;
+    var channelid = this.channelid;
+    var userid = this.userid;
+    var username = this.username;
+    var args = this.argsString; // todo: replace this with slack.args
+    var response_url = this.response_url;
+    var trigger_id = this.trigger_id;
+    
+    // process command field. 
+    // note: Slack has a 3 second timeout on client end. This has not been an issue yet, but with database volume increase, function execution times may increase and lead to timeouts. 
+    // note-continued: A fix would be to return an acknowledgement message to client directly without actually executing the function. 
+    // note-continued: The function should be somehow queued for execution and par.response_url is passed as an extra argument which allows the slack acknowledgment message to be updated. 
+    // note-continued : I have tried delayed function executes with time-delay triggers but that was not appropriate as time-delay triggers - upon creation - have a 1 min queue time before triggering. Too long.
+    // note-continued : A workaround could be to queue commands with an automatic submission to a dedicated form (see https://stackoverflow.com/questions/54809366/how-to-send-delayed-response-slack-api-with-google-apps-script-webapp?rq=1). 
+    // note-continued : The form responses must be linked to the spreadsheet. Then, the onFormSubmit installed trigger may catch the submissions and execute the relevant functions. This is faster than time-delayed triggers, but can still take several seconds. 
+    // note-continued : For now I have avoided any delayed execution and optimised the function execution times to ensure that a return message arrives within the 3 second timeout window.
+    if(command==='done_clarify'){
+      done_process_modal(userid,this.args);
+      return ContentService.createTextOutput(); // an empty HTTP 200 OK message is required for modal to close on slack client end.
+    } else if (command == '/_volunteer'){
+      return volunteer(args, channelid, userid, username);      
```

Comment:
Thinking out loud, it should be possible to make this asynchronous by adding something like:

https://stackoverflow.com/a/49101767

Plus:
```
function runFunctionAsync(functionNameAsString, [args]) {
 
  var trigger = ScriptApp.newTrigger(functionNameAsString).timeBased()
    .after(100)
    .create();

  setupTriggerArguments(trigger, args, false);

    return contentServerJsonReply('Thank you for your message. I\'m a poor bot so please be patient... it should take me up to a few minutes to get back to you...');
}
```

Plus making the output of each function a post request.